File: es63x/src/main/java/tech/beshu/ror/es/RequestInfo.java
/*
 *    This file is part of ReadonlyREST.
 *
 *    ReadonlyREST is free software: you can redistribute it and/or modify
 *    it under the terms of the GNU General Public License as published by
 *    the Free Software Foundation, either version 3 of the License, or
 *    (at your option) any later version.
 *
 *    ReadonlyREST is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public License
 *    along with ReadonlyREST.  If not, see http://www.gnu.org/licenses/
 */

package tech.beshu.ror.es;

import com.google.common.base.Strings;
import com.google.common.collect.Sets;
import org.elasticsearch.action.ActionRequest;
import org.elasticsearch.action.CompositeIndicesRequest;
import org.elasticsearch.action.DocWriteRequest;
import org.elasticsearch.action.IndicesRequest;
import org.elasticsearch.action.admin.cluster.repositories.delete.DeleteRepositoryRequest;
import org.elasticsearch.action.admin.cluster.repositories.get.GetRepositoriesRequest;
import org.elasticsearch.action.admin.cluster.repositories.put.PutRepositoryRequest;
import org.elasticsearch.action.admin.cluster.repositories.verify.VerifyRepositoryRequest;
import org.elasticsearch.action.admin.cluster.snapshots.create.CreateSnapshotRequest;
import org.elasticsearch.action.admin.cluster.snapshots.delete.DeleteSnapshotRequest;
import org.elasticsearch.action.admin.cluster.snapshots.get.GetSnapshotsRequest;
import org.elasticsearch.action.admin.cluster.snapshots.restore.RestoreSnapshotRequest;
import org.elasticsearch.action.admin.cluster.snapshots.status.SnapshotsStatusRequest;
import org.elasticsearch.action.admin.indices.alias.IndicesAliasesRequest;
import org.elasticsearch.action.bulk.BulkRequest;
import org.elasticsearch.action.bulk.BulkShardRequest;
import org.elasticsearch.action.delete.DeleteRequest;
import org.elasticsearch.action.get.MultiGetRequest;
import org.elasticsearch.action.index.IndexRequest;
import org.elasticsearch.action.search.MultiSearchRequest;
import org.elasticsearch.action.search.SearchRequest;
import org.elasticsearch.action.termvectors.MultiTermVectorsRequest;
import org.elasticsearch.action.termvectors.TermVectorsRequest;
import org.elasticsearch.cluster.metadata.AliasOrIndex;
import org.elasticsearch.cluster.metadata.IndexMetaData;
import org.elasticsearch.cluster.metadata.IndexNameExpressionResolver;
import org.elasticsearch.cluster.service.ClusterService;
import org.elasticsearch.common.bytes.BytesReference;
import org.elasticsearch.common.util.ArrayUtils;
import org.elasticsearch.index.Index;
import org.elasticsearch.index.reindex.ReindexRequest;
import org.elasticsearch.rest.RestChannel;
import org.elasticsearch.rest.RestRequest;
import org.elasticsearch.search.builder.SearchSourceBuilder;
import org.elasticsearch.threadpool.ThreadPool;
import org.reflections.ReflectionUtils;
import tech.beshu.ror.commons.shims.es.ESContext;
import tech.beshu.ror.commons.shims.es.LoggerShim;
import tech.beshu.ror.commons.shims.request.RequestInfoShim;
import tech.beshu.ror.commons.utils.RCUtils;
import tech.beshu.ror.commons.utils.ReflecUtils;

import java.lang.reflect.Field;
import java.net.InetSocketAddress;
import java.security.AccessController;
import java.security.PrivilegedAction;
import java.util.Collection;
import java.util.Collections;
import java.util.Iterator;
import java.util.Map;
import java.util.Set;
import java.util.SortedMap;
import java.util.TreeMap;
import java.util.stream.Collectors;

import static tech.beshu.ror.commons.utils.ReflecUtils.extractStringArrayFromPrivateMethod;
import static tech.beshu.ror.commons.utils.ReflecUtils.invokeMethodCached;

public class RequestInfo implements RequestInfoShim {

  private final RestRequest request;
  private final String action;
  private final ActionRequest actionRequest;
  private final String id;
  private final ClusterService clusterService;
  private final Long taskId;
  private final IndexNameExpressionResolver indexResolver;
  private final ThreadPool threadPool;
  private final LoggerShim logger;
  private final RestChannel channel;
  private String content = null;
  private Integer contentLength;
  private ESContext context;

  RequestInfo(
      RestChannel channel, Long taskId, String action, ActionRequest actionRequest,
      ClusterService clusterService, ThreadPool threadPool, ESContext context, IndexNameExpressionResolver indexResolver) {
    this.context = context;
    this.logger = context.logger(getClass());
    this.threadPool = threadPool;
    this.request = channel.request();
    this.channel = channel;
    this.action = action;
    this.actionRequest = actionRequest;
    this.clusterService = clusterService;
    this.indexResolver = indexResolver;
    this.taskId = taskId;
    String tmpID = request.hashCode() + "-" + actionRequest.hashCode();
    if (taskId != null) {
      this.id = tmpID + "#" + taskId;
    }
    else {
      this.id = tmpID;
    }
  }

  @Override
  public String extractType() {
    return actionRequest.getClass().getSimpleName();
  }

  @Override
  public Integer getContentLength() {
    if (contentLength == null) {
      BytesReference cnt = request.content();
      if (cnt == null) {
        contentLength = 0;
      }
      else {
        contentLength = request.content().length();
      }
    }
    return contentLength;
  }

  @Override
  public Set<String> extractIndexMetadata(String index) {
    SortedMap<String, AliasOrIndex> lookup = clusterService.state().metaData().getAliasAndIndexLookup();
    return lookup.get(index).getIndices().stream().map(IndexMetaData::getIndexUUID).collect(Collectors.toSet());
  }

  @Override
  public Long extractTaskId() {
    return taskId;
  }

  public RestChannel getChannel() {
    return channel;
  }

  @Override
  public Integer extractContentLength() {
    if (contentLength == null) {
      BytesReference cnt = request.content();
      if (cnt == null) {
        contentLength = 0;
      }
      else {
        contentLength = request.content().length();
      }
    }
    return contentLength;
  }

  @Override
  public String extractContent() {
    if (content == null) {
      try {
        content = request.content().utf8ToString();
      } catch (Exception e) {
        content = "";
      }
    }
    return content;
  }

  @Override
  public String extractMethod() {
    return request.method().name();
  }

  @Override
  public String extractURI() {
    return request.uri();
  }

  @Override
  public Set<String> extractIndices() {

    String[] indices = new String[0];
    ActionRequest ar = actionRequest;

    // The most common case first
    if (ar instanceof IndexRequest) {
      IndexRequest ir = (IndexRequest) ar;
      indices = ir.indices();
    }

    // CompositeIndicesRequests
    else if (ar instanceof MultiGetRequest) {
      MultiGetRequest cir = (MultiGetRequest) ar;

      for (MultiGetRequest.Item ir : cir.getItems()) {
        indices = ArrayUtils.concat(indices, ir.indices(), String.class);
      }
    }

    else if (ar instanceof MultiSearchRequest) {
      MultiSearchRequest cir = (MultiSearchRequest) ar;

      for (SearchRequest ir : cir.requests()) {
        indices = ArrayUtils.concat(indices, ir.indices(), String.class);
      }
    }

    else if (ar instanceof MultiTermVectorsRequest) {
      MultiTermVectorsRequest cir = (MultiTermVectorsRequest) ar;

      for (TermVectorsRequest ir : cir.getRequests()) {
        indices = ArrayUtils.concat(indices, ir.indices(), String.class);
      }
    }

    else if (ar instanceof BulkRequest) {
      BulkRequest cir = (BulkRequest) ar;

      for (DocWriteRequest<?> ir : cir.requests()) {
        indices = ArrayUtils.concat(indices, ir.indices(), String.class);
      }
    }

    else if (ar instanceof DeleteRequest) {
      DeleteRequest ir = (DeleteRequest) ar;
      indices = ir.indices();
    }

    else if (ar instanceof IndicesAliasesRequest) {
      IndicesAliasesRequest ir = (IndicesAliasesRequest) ar;
      Set<String> indicesSet = ir.getAliasActions().stream().map(x -> Sets.newHashSet(x.indices()))
                                 .flatMap(Collection::stream)
                                 .collect(Collectors.toSet());
      indices = (String[]) indicesSet.toArray();
    }

    // Buggy cases here onwards
    else if (ar instanceof ReindexRequest) {
      // Using reflection otherwise need to create another sub-project
      try {
        SearchRequest sr = (SearchRequest) invokeMethodCached(ar, ar.getClass(), "getSearchRequest");
        IndexRequest ir = (IndexRequest) invokeMethodCached(ar, ar.getClass(), "getDestination");
        indices = ArrayUtils.concat(sr.indices(), ir.indices(), String.class);
      } catch (Exception e) {
        e.printStackTrace();
      }
    }
    else if (ar.getClass().getSimpleName().startsWith("Sql")) {
      // Do noting, we can't do anything about X-Pack SQL queries, as it does not contain indices.
      // #TODO The only way we can filter this kind of request is going Lucene level like "filter" rule.
    }

    else if (ar instanceof CompositeIndicesRequest) {
      logger.error(
          "Found an instance of CompositeIndicesRequest that could not be handled: report this as a bug immediately! "
              + ar.getClass().getSimpleName());
    }

    // Particular case because bug: https://github.com/elastic/elasticsearch/issues/28671
    else if (ar instanceof RestoreSnapshotRequest) {
      RestoreSnapshotRequest rsr = (RestoreSnapshotRequest) ar;
      indices = rsr.indices();
    }

    // Last resort
    else {
      indices = extractStringArrayFromPrivateMethod("indices", ar, context);
      if (indices == null || indices.length == 0) {
        indices = extractStringArrayFromPrivateMethod("index", ar, context);
      }
    }

    if (indices == null) {
      indices = new String[0];
    }

    Set<String> indicesSet = Sets.newHashSet(indices);

    if (logger.isDebugEnabled()) {
      String idxs = String.join(",", indicesSet);
      logger.debug("Discovered indices: " + idxs);
    }

    return indicesSet;
  }

  @Override
  public Set<String> extractSnapshots() {
    if (actionRequest instanceof GetSnapshotsRequest) {
      GetSnapshotsRequest rsr = (GetSnapshotsRequest) actionRequest;
      return Sets.newHashSet(rsr.snapshots());
    }
    else if (actionRequest instanceof CreateSnapshotRequest) {
      CreateSnapshotRequest r = (CreateSnapshotRequest) actionRequest;
      return Sets.newHashSet(r.snapshot());
    }
    else if (actionRequest instanceof DeleteSnapshotRequest) {
      DeleteSnapshotRequest r = (DeleteSnapshotRequest) actionRequest;
      return Sets.newHashSet(r.snapshot());
    }
    else if (actionRequest instanceof RestoreSnapshotRequest) {
      RestoreSnapshotRequest r = (RestoreSnapshotRequest) actionRequest;
      return Sets.newHashSet(r.snapshot());
    }
    else if (actionRequest instanceof SnapshotsStatusRequest) {
      SnapshotsStatusRequest r = (SnapshotsStatusRequest) actionRequest;
      return Sets.newHashSet(r.snapshots());
    }
    return Collections.emptySet();
  }

  @Override
  public void writeSnapshots(Set<String> newSnapshots) {
    // We limit this to read requests, as all the write requests are single-snapshot oriented.
    String[] newSnapshotsA = newSnapshots.toArray(new String[newSnapshots.size()]);
    if (actionRequest instanceof GetSnapshotsRequest) {
      GetSnapshotsRequest rsr = (GetSnapshotsRequest) actionRequest;
      rsr.snapshots(newSnapshotsA);
      return;
    }

    if (actionRequest instanceof SnapshotsStatusRequest) {
      SnapshotsStatusRequest r = (SnapshotsStatusRequest) actionRequest;
      r.snapshots(newSnapshotsA);
      return;
    }
  }

  @Override
  public void writeRepositories(Set<String> newRepositories) {
    // We limit this to read requests, as all the write requests are single-snapshot oriented.
    String[] newRepositoriesA = newRepositories.toArray(new String[newRepositories.size()]);
    if (actionRequest instanceof GetSnapshotsRequest) {
      GetSnapshotsRequest rsr = (GetSnapshotsRequest) actionRequest;
      rsr.repository(newRepositoriesA[0]);
      return;
    }

    if (actionRequest instanceof SnapshotsStatusRequest) {
      SnapshotsStatusRequest r = (SnapshotsStatusRequest) actionRequest;
      r.repository(newRepositoriesA[0]);
      return;
    }
    if (actionRequest instanceof GetRepositoriesRequest) {
      GetRepositoriesRequest r = (GetRepositoriesRequest) actionRequest;
      r.repositories(newRepositoriesA);
      return;
    }

    if (actionRequest instanceof VerifyRepositoryRequest) {
      VerifyRepositoryRequest r = (VerifyRepositoryRequest) actionRequest;
      r.name(newRepositoriesA[0]);
      return;
    }

  }
  //  public Set<String> extractAllRepositories() {
  //    RepositoriesMetaData out = clusterService.state().metaData().custom(RepositoriesMetaData.TYPE);
  //    List<RepositoryMetaData> x = out.repositories();
  //    return x.stream().map(RepositoryMetaData::name).collect(Collectors.toSet());
  //    //    clusterService.state().getMetaData().
  //  }

  @Override
  public Set<String> extractRepositories() {
    if (actionRequest instanceof GetSnapshotsRequest) {
      GetSnapshotsRequest rsr = (GetSnapshotsRequest) actionRequest;
      return Sets.newHashSet(rsr.repository());
    }
    else if (actionRequest instanceof CreateSnapshotRequest) {
      CreateSnapshotRequest r = (CreateSnapshotRequest) actionRequest;
      return Sets.newHashSet(r.repository());
    }
    else if (actionRequest instanceof DeleteSnapshotRequest) {
      DeleteSnapshotRequest r = (DeleteSnapshotRequest) actionRequest;
      return Sets.newHashSet(r.repository());
    }
    else if (actionRequest instanceof RestoreSnapshotRequest) {
      RestoreSnapshotRequest r = (RestoreSnapshotRequest) actionRequest;
      return Sets.newHashSet(r.repository());
    }
    else if (actionRequest instanceof SnapshotsStatusRequest) {
      SnapshotsStatusRequest r = (SnapshotsStatusRequest) actionRequest;
      return Sets.newHashSet(r.repository());
    }

    // Specific to repositories
    else if (actionRequest instanceof PutRepositoryRequest) {
      PutRepositoryRequest r = (PutRepositoryRequest) actionRequest;
      return Sets.newHashSet(r.name());
    }
    else if (actionRequest instanceof GetRepositoriesRequest) {
      GetRepositoriesRequest r = (GetRepositoriesRequest) actionRequest;
      return Sets.newHashSet(r.repositories());
    }
    else if (actionRequest instanceof DeleteRepositoryRequest) {
      DeleteRepositoryRequest r = (DeleteRepositoryRequest) actionRequest;
      return Sets.newHashSet(r.name());
    }
    else if (actionRequest instanceof VerifyRepositoryRequest) {
      VerifyRepositoryRequest r = (VerifyRepositoryRequest) actionRequest;
      return Sets.newHashSet(r.name());
    }

    return Collections.emptySet();
  }

  @Override
  public String extractAction() {
    return action;
  }

  @Override
  public String extractId() {
    return id;
  }

  @Override
  public Map<String, String> extractRequestHeaders() {
    final Map<String, String> h = new TreeMap<>(String.CASE_INSENSITIVE_ORDER);
    request.getHeaders().keySet().forEach(k -> {
      if (request.getAllHeaderValues(k).isEmpty()) {
        return;
      }
      h.put(k, request.getAllHeaderValues(k).iterator().next());
    });
    return h;
  }

  @Override
  public String extractLocalAddress() {
    String remoteHost = ((InetSocketAddress) request.getLocalAddress()).getAddress().getHostAddress();
    return remoteHost;
  }

  @Override
  public String extractRemoteAddress() {
    String remoteHost = ((InetSocketAddress) request.getRemoteAddress()).getAddress().getHostAddress();
    // Make sure we recognize localhost even when IPV6 is involved
    if (RCUtils.isLocalHost(remoteHost)) {
      remoteHost = RCUtils.LOCALHOST;
    }
    return remoteHost;
  }

  @Override
  public void writeIndices(Set<String> newIndices) {
    // Setting indices by reflection..
    newIndices.remove("<no-index>");
    newIndices.remove("");

    // Best case, this request is designed to have indices replaced.
    if (actionRequest instanceof IndicesRequest.Replaceable) {
      ((IndicesRequest.Replaceable) actionRequest).indices(newIndices.toArray(new String[newIndices.size()]));
      return;
    }

    // This should not be necessary anymore because nowadays we either allow or forbid write requests.
    if (actionRequest instanceof BulkShardRequest) {
      BulkShardRequest bsr = (BulkShardRequest) actionRequest;
      String singleIndex = newIndices.iterator().next();
      String uuid = extractIndexMetadata(singleIndex).iterator().next();
      AccessController.doPrivileged((PrivilegedAction<Void>) () -> {
        @SuppressWarnings("unchecked")
        Set<Field> fields = ReflectionUtils.getAllFields(bsr.shardId().getClass(), ReflectionUtils.withName("index"));
        fields.stream().forEach(f -> {
          f.setAccessible(true);
          try {
            f.set(bsr.shardId(), new Index(singleIndex, uuid));
          } catch (Throwable e) {
            e.printStackTrace();
          }
        });
        return null;
      });
    }

    if (actionRequest instanceof MultiSearchRequest) {
      // If it's an empty MSR, we are ok
      MultiSearchRequest msr = (MultiSearchRequest) actionRequest;
      for (SearchRequest sr : msr.requests()) {

        // This contains global indices
        if (sr.indices().length == 0 || Sets.newHashSet(sr.indices()).contains("*")) {
          sr.indices(newIndices.toArray(new String[newIndices.size()]));
          continue;
        }

        // This transforms wildcards and aliases in concrete indices
        Set<String> expandedSrIndices = getExpandedIndices(Sets.newHashSet(sr.indices()));

        Set<String> remaining = Sets.newHashSet(expandedSrIndices);
        remaining.retainAll(newIndices);

        if (remaining.size() == 0) {
          // contained just forbidden indices, should return zero results
          sr.source(new SearchSourceBuilder().size(0));
          continue;
        }
        if (remaining.size() == expandedSrIndices.size()) {
          // contained all allowed indices
          continue;
        }
        // some allowed indices were there, restrict query to those
        sr.indices(remaining.toArray(new String[remaining.size()]));
      }
      // All the work is done - no need for reflection
      return;
    }

    if (actionRequest instanceof MultiGetRequest) {
      MultiGetRequest mgr = (MultiGetRequest) actionRequest;
      Iterator<MultiGetRequest.Item> it = mgr.getItems().iterator();
      while (it.hasNext()) {
        MultiGetRequest.Item item = it.next();
        // One item contains just an index, but can be an alias
        Set<String> indices = getExpandedIndices(Sets.newHashSet(item.indices()));
        Set<String> remaining = indices;
        remaining.retainAll(newIndices);
        if (remaining.isEmpty()) {
          it.remove();
        }
      }
      // All the work is done - no need for reflection
      return;
    }

    if (actionRequest instanceof IndicesAliasesRequest) {
      IndicesAliasesRequest iar = (IndicesAliasesRequest) actionRequest;
      Iterator<IndicesAliasesRequest.AliasActions> it = iar.getAliasActions().iterator();
      while (it.hasNext()) {
        IndicesAliasesRequest.AliasActions act = it.next();
        Set<String> indices = getExpandedIndices(Sets.newHashSet(act.indices()));
        Set<String> remaining = indices;
        remaining.retainAll(newIndices);
        if (remaining.isEmpty()) {
          it.remove();
          continue;
        }
        act.indices(remaining.toArray(new String[remaining.size()]));
      }
      // All the work is done - no need for reflection
      return;
    }

    // Optimistic reflection attempt
    boolean okSetResult = ReflecUtils.setIndices(actionRequest, Sets.newHashSet("index", "indices"), newIndices, logger);

    if (okSetResult) {
      if (logger.isDebugEnabled()) {
        logger.debug("REFLECTION: success changing indices: " + newIndices + " correctly set as " + extractIndices());
      }
    }
    else {
      logger.error("REFLECTION: Failed to set indices for type " + actionRequest.getClass().getSimpleName() +
          "  in req id: " + extractId());
    }
  }

  @Override
  public void writeResponseHeaders(Map<String, String> hMap) {
    hMap.keySet().forEach(k -> threadPool.getThreadContext().addResponseHeader(k, hMap.get(k)));
  }

  @Override
  public Set<String> extractAllIndicesAndAliases() {
    return clusterService.state().metaData().getAliasAndIndexLookup().keySet();
  }

  @Override
  public boolean involvesIndices() {
    return actionRequest instanceof IndicesRequest ||
        actionRequest instanceof CompositeIndicesRequest ||
        // Necessary because it won't implement IndicesRequest as it should (bug: https://github.com/elastic/elasticsearch/issues/28671)
        actionRequest instanceof RestoreSnapshotRequest;

  }

  @Override
  public boolean extractIsReadRequest() {
    return RCUtils.isReadRequest(action);
  }

  @Override
  public boolean extractIsCompositeRequest() {
    return actionRequest instanceof CompositeIndicesRequest;
  }

  @Override
  public void writeToThreadContextHeader(String key, String value) {
    threadPool.getThreadContext().putHeader(key, value);
  }

  @Override
  public String consumeThreadContextHeader(String key) {
    String value = threadPool.getThreadContext().getHeader(key);
    if (!Strings.isNullOrEmpty(value)) {
      threadPool.getThreadContext().getHeaders().remove(key);
    }
    return value;
  }

}

File: es73x/src/main/scala/tech/beshu/ror/es/RequestInfo.scala
/*
 *    This file is part of ReadonlyREST.
 *
 *    ReadonlyREST is free software: you can redistribute it and/or modify
 *    it under the terms of the GNU General Public License as published by
 *    the Free Software Foundation, either version 3 of the License, or
 *    (at your option) any later version.
 *
 *    ReadonlyREST is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public License
 *    along with ReadonlyREST.  If not, see http://www.gnu.org/licenses/
 */
package tech.beshu.ror.es

import java.util.UUID
import java.{lang, util}

import com.google.common.collect.{Maps, Sets}
import org.apache.logging.log4j.scala.Logging
import org.elasticsearch.action.admin.cluster.repositories.delete.DeleteRepositoryRequest
import org.elasticsearch.action.admin.cluster.repositories.get.GetRepositoriesRequest
import org.elasticsearch.action.admin.cluster.repositories.put.PutRepositoryRequest
import org.elasticsearch.action.admin.cluster.repositories.verify.VerifyRepositoryRequest
import org.elasticsearch.action.admin.cluster.snapshots.create.CreateSnapshotRequest
import org.elasticsearch.action.admin.cluster.snapshots.delete.DeleteSnapshotRequest
import org.elasticsearch.action.admin.cluster.snapshots.get.GetSnapshotsRequest
import org.elasticsearch.action.admin.cluster.snapshots.restore.RestoreSnapshotRequest
import org.elasticsearch.action.admin.cluster.snapshots.status.SnapshotsStatusRequest
import org.elasticsearch.action.admin.indices.alias.IndicesAliasesRequest
import org.elasticsearch.action.admin.indices.template.delete.DeleteIndexTemplateRequest
import org.elasticsearch.action.admin.indices.template.get.GetIndexTemplatesRequest
import org.elasticsearch.action.admin.indices.template.put.PutIndexTemplateRequest
import org.elasticsearch.action.bulk.{BulkRequest, BulkShardRequest}
import org.elasticsearch.action.delete.DeleteRequest
import org.elasticsearch.action.get.MultiGetRequest
import org.elasticsearch.action.index.IndexRequest
import org.elasticsearch.action.search.{MultiSearchRequest, SearchRequest}
import org.elasticsearch.action.termvectors.MultiTermVectorsRequest
import org.elasticsearch.action.{ActionRequest, CompositeIndicesRequest, IndicesRequest}
import org.elasticsearch.cluster.service.ClusterService
import org.elasticsearch.index.Index
import org.elasticsearch.index.reindex.ReindexRequest
import org.elasticsearch.rest.RestChannel
import org.elasticsearch.search.builder.SearchSourceBuilder
import org.elasticsearch.threadpool.ThreadPool
import org.elasticsearch.transport.RemoteClusterService
import org.reflections.ReflectionUtils
import tech.beshu.ror.es.utils.ClusterServiceHelper.{findTemplatesOfIndices, getIndicesRelatedToTemplates, indicesFromPatterns}
import tech.beshu.ror.shims.request.RequestInfoShim
import tech.beshu.ror.utils.LoggerOps._
import tech.beshu.ror.utils.ReflecUtils.{extractStringArrayFromPrivateMethod, invokeMethodCached}
import tech.beshu.ror.utils.{RCUtils, ReflecUtils}

import scala.collection.JavaConverters._
import scala.math.Ordering.comparatorToOrdering
import scala.util.{Failure, Success, Try}

class RequestInfo(channel: RestChannel, taskId: lang.Long, action: String, actionRequest: ActionRequest,
                  clusterService: ClusterService, threadPool: ThreadPool, remoteClusterService: RemoteClusterService)
  extends RequestInfoShim with Logging {

  private val request = channel.request()

  override def extractType(): String = actionRequest.getClass.getSimpleName

  override def extractIndexMetadata(index: String): util.Set[String] = {
    val lookup = clusterService.state.metaData.getAliasAndIndexLookup
    lookup.get(index).getIndices.asScala.map(_.getIndexUUID).toSet.asJava
  }

  override val extractTaskId: lang.Long = taskId

  override lazy val extractContentLength: Integer = if(request.content == null) 0 else request.content().length()

  override lazy val extractContent: String = if(request.content == null) "" else request.content().utf8ToString()

  override lazy val extractMethod: String = request.method().name()

  override val extractURI: String = request.uri()

  override val involvesIndices: Boolean = {
    actionRequest.isInstanceOf[IndicesRequest] || actionRequest.isInstanceOf[CompositeIndicesRequest] ||
      // Necessary because it won't implement IndicesRequest as it should (bug: https://github.com/elastic/elasticsearch/issues/28671)
      actionRequest.isInstanceOf[RestoreSnapshotRequest] ||
      actionRequest.isInstanceOf[GetIndexTemplatesRequest] || actionRequest.isInstanceOf[PutIndexTemplateRequest] || actionRequest.isInstanceOf[DeleteIndexTemplateRequest]
  }

  override lazy val extractIndices: util.Set[String] = {
    val indices = actionRequest match {
      case ar: IndexRequest => // The most common case first
        ar.indices.toSet
      case ar: MultiGetRequest =>
        ar.getItems.asScala.flatMap(_.indices()).toSet
      case ar: MultiSearchRequest =>
        ar.requests().asScala.flatMap(_.indices()).toSet
      case ar: MultiTermVectorsRequest =>
        ar.getRequests.asScala.flatMap(_.indices()).toSet
      case ar: BulkRequest =>
        ar.requests().asScala.flatMap(_.indices()).toSet
      case ar: DeleteRequest =>
        ar.indices().toSet
      case ar: IndicesAliasesRequest =>
        ar.getAliasActions.asScala.flatMap(_.indices()).toSet
      case ar: ReindexRequest => // Buggy cases here onwards
        Try {
          val sr = invokeMethodCached(ar, ar.getClass, "getSearchRequest").asInstanceOf[SearchRequest]
          val ir = invokeMethodCached(ar, ar.getClass, "getDestination").asInstanceOf[IndexRequest]
          sr.indices().toSet ++ ir.indices().toSet
        } fold(
          ex => {
            logger.errorEx(s"cannot extract indices from: $extractMethod $extractURI\n$extractContent", ex)
            Set.empty[String]
          },
          identity
        )
      case ar if ar.getClass.getSimpleName.startsWith("Sql") =>
        // Do noting, we can't do anything about X-Pack SQL queries, as it does not contain indices.
        // todo: The only way we can filter this kind of request is going Lucene level like "filter" rule.
        Set.empty[String]
      case ar: CompositeIndicesRequest =>
        logger.error(s"Found an instance of CompositeIndicesRequest that could not be handled: report this as a bug immediately! ${ar.getClass.getSimpleName}")
        Set.empty[String]
      case ar: RestoreSnapshotRequest => // Particular case because bug: https://github.com/elastic/elasticsearch/issues/28671
        ar.indices().toSet
      case ar: PutIndexTemplateRequest =>
        indicesFromPatterns(clusterService, ar.indices.toSet)
      case ar: DeleteIndexTemplateRequest =>
        getIndicesRelatedToTemplates(clusterService, Set(ar.name))
      case ar: GetIndexTemplatesRequest =>
        if(ar.names().isEmpty) Set("*")
        else getIndicesRelatedToTemplates(clusterService, ar.names().toSet)
      case ar =>
        val indices = extractStringArrayFromPrivateMethod("indices", ar).toSet
        if(indices.isEmpty) extractStringArrayFromPrivateMethod("index", ar).toSet
        else indices
    }
    logger.debug(s"Discovered indices: ${indices.mkString(",")}")
    indices.asJava
  }

  override lazy val extractSnapshots: util.Set[String] = {
    actionRequest match {
      case ar: GetSnapshotsRequest => ar.snapshots().toSet.asJava
      case ar: CreateSnapshotRequest => Set(ar.snapshot()).asJava
      case ar: DeleteSnapshotRequest => Set(ar.snapshot()).asJava
      case ar: RestoreSnapshotRequest => Set(ar.snapshot()).asJava
      case ar: SnapshotsStatusRequest => ar.snapshots().toSet.asJava
      case _ => Set.empty[String].asJava
    }
  }

  override lazy val extractRepositories: util.Set[String] = {
    actionRequest match {
      case ar: GetSnapshotsRequest => Set(ar.repository()).asJava
      case ar: CreateSnapshotRequest => Set(ar.repository()).asJava
      case ar: DeleteSnapshotRequest => Set(ar.repository()).asJava
      case ar: RestoreSnapshotRequest => Set(ar.repository()).asJava
      case ar: SnapshotsStatusRequest => Set(ar.repository()).asJava
      case ar: PutRepositoryRequest => Set(ar.name()).asJava
      case ar: GetRepositoriesRequest => ar.repositories().toSet.asJava
      case ar: DeleteRepositoryRequest => Set(ar.name()).asJava
      case ar: VerifyRepositoryRequest => Set(ar.name()).asJava
      case _ => Set.empty[String].asJava
    }
  }

  override val extractAction: String = action

  override val extractRequestHeaders: util.Map[String, String] =
    request
      .getHeaders.asScala
      .map { h => (h._1, h._2.asScala.min(comparatorToOrdering(String.CASE_INSENSITIVE_ORDER))) }
      .asJava

  override val extractRemoteAddress: String = {
    Try {
      val remoteHost = request.getHttpChannel.getRemoteAddress.getAddress.getHostAddress
      // Make sure we recognize localhost even when IPV6 is involved
      if (RCUtils.isLocalHost(remoteHost)) RCUtils.LOCALHOST else remoteHost
    } fold(
      ex => {
        logger.error("Could not extract remote address", ex)
        null
      },
      identity
    )
  }

  override val extractLocalAddress: String = {
    Try {
      request.getHttpChannel.getLocalAddress.getAddress.getHostAddress
    } fold(
      ex => {
        logger.error("Could not extract local address", ex)
        null
      },
      identity
    )
  }

  override val extractId: String = {
    val tmpID = request.hashCode() + "-" + actionRequest.hashCode()
    if (taskId != null) s"$tmpID#$taskId"
    else tmpID
  }

  override lazy val extractAllIndicesAndAliases: util.Set[java.util.Map.Entry[String, util.Set[String]]] = {
    val indices = clusterService.state.metaData.getIndices
    indices
      .keysIt().asScala
      .map { index =>
        val indexMetaData = indices.get(index)
        val indexName = indexMetaData.getIndex.getName
        val aliases: util.Set[String] = Sets.newHashSet(indexMetaData.getAliases.keysIt)
        Maps.immutableEntry(indexName, aliases)
      }
      .toSet
      .asJava
  }

  override val extractIsReadRequest: Boolean = RCUtils.isReadRequest(action)

  override val extractIsAllowedForDLS: Boolean = {
    actionRequest match {
      case _ if !extractIsReadRequest => false
      case sr: SearchRequest if sr.source() == null => true
      case sr: SearchRequest if sr.source.profile || (sr.source.suggest != null && !sr.source.suggest.getSuggestions.isEmpty) => false
      case _ => true
    }
  }

  override val extractIsCompositeRequest: Boolean = actionRequest.isInstanceOf[CompositeIndicesRequest]

  override val extractHasRemoteClusters: Boolean = remoteClusterService.isCrossClusterSearchEnabled

  override def writeSnapshots(newSnapshots: util.Set[String]): Unit = {
    if (newSnapshots.isEmpty) return

    // We limit this to read requests, as all the write requests are single-snapshot oriented.
    actionRequest match {
      case ar: GetSnapshotsRequest =>
        ar.snapshots(newSnapshots.toArray(new Array[String](newSnapshots.size)))
      case ar: SnapshotsStatusRequest =>
        ar.snapshots(newSnapshots.toArray(new Array[String](newSnapshots.size)))
      case _ =>
    }
  }

  override def writeRepositories(newRepositories: util.Set[String]): Unit = {
    if (newRepositories.isEmpty) return

    // We limit this to read requests, as all the write requests are single-snapshot oriented.
    val newRepositoriesA = newRepositories.toArray(new Array[String](newRepositories.size))
    actionRequest match {
      case ar: GetSnapshotsRequest => ar.repository(newRepositoriesA(0))
      case ar: SnapshotsStatusRequest => ar.repository(newRepositoriesA(0))
      case ar: GetRepositoriesRequest => ar.repositories(newRepositoriesA)
      case ar: VerifyRepositoryRequest => ar.name(newRepositoriesA(0))
      case _ =>
    }
  }

  override def writeResponseHeaders(hMap: util.Map[String, String]): Unit = {
    val threadContext = threadPool.getThreadContext
    hMap.asScala.foreach { case (key, value) =>
      threadContext.addResponseHeader(key, value)
    }
  }

  override def writeToThreadContextHeaders(hMap: util.Map[String, String]): Unit = {
    val threadContext = threadPool.getThreadContext
    hMap.asScala.foreach { case (key, value) =>
      threadContext.putHeader(key, value)
    }
  }

  override def writeIndices(newIndices: util.Set[String]): Unit = {
    val indices = newIndices.asScala.filter(i => i != "" && i != "<no-index>").toList
    if (indices.isEmpty) return

    actionRequest match {
      case ar: IndicesRequest.Replaceable => // Best case, this request is designed to have indices replaced.
        ar.indices(indices: _*)
      case ar: BulkShardRequest => // This should not be necessary anymore because nowadays we either allow or forbid write requests.
        val singleIndex = indices.head
        val uuid = extractIndexMetadata(singleIndex).asScala.toList.head
        ReflectionUtils
          .getAllFields(ar.shardId().getClass, ReflectionUtils.withName("index")).asScala
          .foreach { f =>
            f.setAccessible(true)
            Try(f.set(ar.shardId(), new Index(singleIndex, uuid))) match {
              case Failure(ex) => logger.error("Cannot set index", ex)
              case Success(_) =>
            }
          }
      case ar: MultiSearchRequest =>
        ar.requests().asScala.foreach { sr =>
          if (sr.indices.length == 0 || sr.indices().contains("*")) {
            sr.indices(indices: _*)
          } else {
            // This transforms wildcards and aliases in concrete indices
            val expandedSrIndices = getExpandedIndices(sr.indices.toSet.asJava)

            val remaining = Sets.newHashSet(expandedSrIndices)
            remaining.retainAll(indices.asJava)

            if (remaining.size == 0) { // contained just forbidden indices, should return zero results
              sr.source(new SearchSourceBuilder().size(0))
            } else if (remaining.size == expandedSrIndices.size) { // contained all allowed indices
              // nothing to do
            } else {
              // some allowed indices were there, restrict query to those
              sr.indices(remaining.asScala.toList: _*)
            }
          }
        }
      case ar: MultiGetRequest =>
        val it = ar.getItems.iterator
        while (it.hasNext) {
          val item = it.next
          // One item contains just an index, but can be an alias
          val indices = getExpandedIndices(item.indices.toSet.asJava)
          val remaining = indices
          remaining.retainAll(indices)
          if (remaining.isEmpty) it.remove()
        }
      case ar: IndicesAliasesRequest =>
        val it = ar.getAliasActions.iterator
        while (it.hasNext) {
          val act = it.next
          val indices = getExpandedIndices(act.indices().toSet.asJava)
          val remaining = indices
          remaining.retainAll(indices)
          if (remaining.isEmpty) {
            it.remove()
          } else {
            act.indices(remaining.asScala.toList: _*)
          }
        }
      case ar: GetIndexTemplatesRequest =>
        val requestTemplateNames = ar.names().toSet
        val allowedTemplateNames = findTemplatesOfIndices(clusterService, indices.toSet)
        val templateNamesToReturn =
          if (requestTemplateNames.isEmpty) allowedTemplateNames
          else requestTemplateNames.intersect(allowedTemplateNames)
        if (templateNamesToReturn.isEmpty) { // hack! there is no other way to return empty list of templates
          ar.names(UUID.randomUUID + "*")
        } else {
          ar.names(templateNamesToReturn.toList: _*)
        }
      case _ =>
        // Optimistic reflection attempt
        val okSetResult = ReflecUtils.setIndices(actionRequest, Sets.newHashSet("index", "indices"), indices.toSet.asJava)
        if (okSetResult) logger.debug(s"REFLECTION: success changing indices: $indices correctly set as $extractIndices")
        else logger.error(s"REFLECTION: Failed to set indices for type ${actionRequest.getClass.getSimpleName} in req id: $extractId")
    }
  }
}
